packages/java/parser-jvm-core/src/main/java/com/vaadin/fusion/parser/core/Parser.java

```diff
@@ -0,0 +1,117 @@
+                });
+    }
+
+    public OpenAPI execute() {
+        logger.debug("Executing JVM Parser");
+        var pluginManager = new PluginManager(config);
+
+        var classPathElements = config.getClassPathElements();
+        logger.debug("Scanning JVM classpath: "
+                + String.join(";", classPathElements));
+        var result = new ClassGraph().enableAllInfo()
+                .overrideClasspath(classPathElements).scan();
+
+        var collector = new EntitiesCollector(result, logger);
+
+        logger.debug(
+                "Checking if the compiler is run with -parameters option enabled");
+        checkIfJavaCompilerParametersFlagIsEnabled(collector.getEndpoints());
+
+        logger.debug("Executing parser plugins");
+        pluginManager.execute(collector.getEndpoints(), collector.getEntities(),
+                storage);
+
+        logger.debug("Parsing process successfully finished");
+        return storage.getOpenAPI();
+    }
+
+    @Nonnull
+    public SharedStorage getStorage() {
+        return storage;
+    }
+
+    private class EntitiesCollector {
+        private final List<RelativeClassInfo> endpoints;
+        private final List<RelativeClassInfo> entities;
+
+        public EntitiesCollector(ScanResult result, Logger logger) {
+            var endpointAnnotationName = config.getEndpointAnnotationName();
+
+            logger.debug(
+                    "Collecting project endpoints with the endpoint annotation: "
+                            + endpointAnnotationName);
+
+            endpoints = result.getClassesWithAnnotation(endpointAnnotationName)
+                    .stream().map(RelativeClassInfo::new)
+                    .collect(Collectors.toList());
+
+            logger.debug("Collected project endpoints: " + endpoints.stream()
+                    .map(RelativeClassInfo::get).map(ClassInfo::getName)
+                    .collect(Collectors.joining(", ")));
+
+            entities = endpoints.stream().flatMap(
+                    cls -> cls.getInheritanceChain().getDependenciesStream())
+                    .distinct().collect(Collectors.toList());
+
+            // ATTENTION: This loop mutates the collection during processing!
+            // It is necessary to collect all endpoint dependencies +
+            // dependencies of dependencies.
+            // Be careful changing it: the endless loop could happen.
+            for (var i = 0; i < entities.size(); i++) {
+                var entity = entities.get(i);
+
+                entity.getDependenciesStream()
+                        .filter(e -> !entities.contains(e))
+                        .forEach(entities::add);
+            }
+
+            logger.debug("Collected project data entities: " + entities.stream()
+                    .map(RelativeClassInfo::get).map(ClassInfo::getName)
+                    .collect(Collectors.joining(", ")));
+        }
+
+        public List<RelativeClassInfo> getEndpoints() {
```

Comment:
```suggestion
        public Collection<RelativeClassInfo> getEndpoints() {
```

considering we might want to use a `LinkedHashSet`, maybe better to go with a more generic API here and down the road in the `Plugin` parameters.